chore(host): Replace anyhow with thiserror

Levi0804 opened the pull request:
This PR is a work in progress addressing #916.

I am aware that some of the changes in this PR introduce breaking changes in other parts of the workspace. I created this PR to take reviews on whether my approach looks good before proceeding further. Please let me know if these changes look good to you.

mattsse commented on bin/host/src/server.rs:
```diff
@@ -20,6 +19,16 @@ pub struct PreimageServer<P, H, B> {
     backend: Arc<B>,
 }
 
+#[derive(Debug, thiserror::Error)]
+enum PreimageServerError {
+    #[error("Failed to serve preimage request: {0}")]
+    PreimageRequestFailed(PreimageOracleError),
+    #[error("Failed to route hint: {0}")]
+    RouteHintFailed(PreimageOracleError),
+    #[error("Join error: {0}")]
+    ExecutionError(#[from] tokio::task::JoinError),
+}
```
this is used it pub api and should also be pub

a one-line doc also wouldn't hurt here

refcell replied:
Agree with @mattsse here

mattsse reviewed (changes requested):
smol style nit

Levi0804 commented:
Will it be okay to allow `unnameable_types` in this case?

refcell commented on bin/host/src/interop/cfg.rs:
```diff
@@ -98,9 +97,28 @@ pub struct InteropHost {
 }
 
+/// An error that can occur when handling interop hosts
+#[derive(Debug, thiserror::Error)]
```
Add [`derive_more::From`](https://docs.rs/derive_more/latest/derive_more/derive.From.html) here - see comments below.

refcell commented on bin/host/src/server.rs:
```diff
@@ -20,6 +19,17 @@ pub struct PreimageServer<P, H, B> {
 
+/// An error that can occur when handling preimage requests
+#[derive(Debug, thiserror::Error)]
+pub enum PreimageServerError {
```
This will cause ci to error because the type is only visible within the `server` module. It has to be visible at the top-level of the crate so linting will pass.

If you run `just l` locally which runs linting, you'll see a visibility error.

The solution is in the host's `lib.rs` file, make the `PreimageServerError` visible by adding it to the `pub use server::` line [here](https://github.com/op-rs/kona/blob/main/bin/host/src/lib.rs#L5).

You'll also need to do this for the other error types introduced in this PR.

refcell commented on bin/host/src/server.rs:
```diff
@@ -20,6 +19,17 @@ pub struct PreimageServer<P, H, B> {
+/// An error that can occur when handling preimage requests
+#[derive(Debug, thiserror::Error)]
+pub enum PreimageServerError {
+    #[error("Failed to serve preimage request: {0}")]
+    PreimageRequestFailed(PreimageOracleError),
+    #[error("Failed to route hint: {0}")]
+    RouteHintFailed(PreimageOracleError),
+    #[error("Join error: {0}")]
+    ExecutionError(#[from] tokio::task::JoinError),
```
Our linting rules are fairly strict as well so these enum variants also need documentation.

refcell commented on bin/host/src/single/cfg.rs:
```diff
@@ -109,9 +108,24 @@ pub struct SingleChainHost {
+/// An error that can occur when handling single chain hosts
+#[derive(Debug, thiserror::Error)]
+pub enum SingleChainHostError {
+    #[error("Error handling preimage request: {0}")]
+    PreimageServerError(#[from] PreimageServerError),
+    #[error("IO error: {0}")]
+    IOError(#[from] std::io::Error),
+    #[error("Failed deserializing RollupConfig: {0}")]
+    ParseError(#[from] serde_json::Error),
+    #[error("Join error: {0}")]
+    ExecutionError(#[from] tokio::task::JoinError),
+    #[error("Error: {0}")]
+    Other(&'static str),
```
These enum variants will need docs

refcell commented on bin/host/src/interop/cfg.rs:
```diff
@@ -98,9 +97,28 @@ pub struct InteropHost {
+/// An error that can occur when handling interop hosts
+#[derive(Debug, thiserror::Error)]
+pub enum InteropHostError {
+    #[error("Error handling preimage request: {0}")]
+    PreimageServerError(#[from] PreimageServerError),
+    #[error("IO error: {0}")]
+    IOError(#[from] std::io::Error),
+    #[error("Failed deserializing RollupConfig: {0}")]
+    ParseError(#[from] serde_json::Error),
+    #[error("Join error: {0}")]
+    ExecutionError(#[from] tokio::task::JoinError),
+    #[error("Rpc Error: {0}")]
+    RcpError(#[from] alloy_transport::RpcError<alloy_transport::TransportErrorKind>),
+    #[error("No provider found for chain ID: {0}")]
+    RootProviderError(u64),
+    #[error("Error: {0}")]
+    Other(&'static str),
```
These enum variants will need docs

refcell commented:
Few more comments for CI, otherwise this is looking good. Thanks for taking this on!

Levi0804 commented:
> Few more comments for CI, otherwise this is looking good. Thanks for taking this on!

Happy to contribute! Let me know if these changes look good. I squashed and rebased to clean up some mistakes...

refcell reviewed (approved):
Nice work!
